feat: Export plotly-express as a dashboard plugin

vbabich opened the pull request:
Tested with DHE V+ (1.20231218.176) Core 0.32.1
Closes #308


BREAKING CHANGE:
- `widget` type in the `PanelEvent.OPEN` event arguments has changed from `VariableDefinition` to `VariableDescriptor`

mofojed commented on plugins/plotly-express/src/js/src/index.ts:
```diff
@@ -1,6 +1,3 @@
-import { PlotlyExpressPlugin } from './PlotlyExpressPlugin.js';
-
+export * from './DashboardPlugin.js';
```
I _think_ you could just add `DashboardPlugin` to `PlotlyExpressPlugin`, and then Enterprise should pick it up as a LegacyDashboardPlugin and load it like that.

Will need the `PlotlyExpressPlugin` type to extend `WidgetPlugin`, e.g.
```

export const PlotlyExpressPlugin: WidgetPlugin<Widget> & { DashboardPlugin: ... } = {
  name: '@deephaven/plotly-express',
  type: PluginType.WIDGET_PLUGIN,
  DashboardPlugin,
  supportedTypes: 'deephaven.plot.express.DeephavenFigure',
  component: PlotlyExpressChart,
  panelComponent: PlotlyExpressChartPanel,
  icon: vsGraph,
};
```

vbabich replied:
I just added `DashboardPlugin` as a named export next to the `default PlotlyExpressPlugin`, didn't have to extend the `WidgetPlugin`.

mattrunyon commented:
I think this will break plotly-express in DHC w/ dh.ui. We currently check if the plugin is the legacy format first when loading, so the legacy definition (named `DashboardPlugin` export) will override the preferred definition (default export of the plugin object)

mattrunyon commented on plugins/plotly-express/src/js/src/index.ts:
```diff
@@ -1,6 +1,3 @@
 export * from './PlotlyExpressChartModel.js';
 export * from './PlotlyExpressChartUtils.js';
-
-export default PlotlyExpressPlugin;
```
The default export needs to stay so DHC can use the new plugin format still

The new format looks at the default export and uses the `type` field of the default to determine what kind of plugin it is. The old format used specifically named exports to differentiate what type of plugin it was

vbabich commented:
I added a named export for the legacy `DashboardPlugin` next to the `default` `WidgetPlugin`.
Updated DHE to prioritize legacy plugins until we implement `WidgetLoaderPlugin` in Enterprise.

mofojed commented on plugins/plotly-express/src/js/src/DashboardPlugin.tsx:
```diff
@@ -0,0 +1,78 @@
+          id: panelId,
+          metadata: {
+            ...metadata,
+            name,
```
Spread the whole widget descriptor here too, `...widget`

vbabich replied:
Done. Should I mark this PR as a breaking change since the `widget` argument type has changed from `VariableDefinition` to `VariableDescriptor`?

mofojed replied:
Yea - we should be sure to be emitting `VariableDescriptor` instead of `VariableDefinition`. I think I've already taken care of that on Enterprise though.

mofojed commented on plugins/plotly-express/src/js/src/DashboardPlugin.tsx:
```diff
@@ -0,0 +1,77 @@
+): JSX.Element | null {
+  const { id, layout, registerComponent } = props;
+
+  const handlePanelOpen = useCallback(
```
Side note, we have a `useDashboardPanel` hook that should do what we're doing here, but it would require other changes as well (since it doesn't use `metadata` or `fetch` at all). No need to worry about it here.

mattrunyon reviewed (approved):
I think we'll actually want DHE to register both plugins. Otherwise plotly-express won't work in dh.ui in DHE

vbabich commented:
Why? What plugin type is needed for dh.ui in DHE?

mattrunyon commented:
dh.ui reads the plugin registry and looks for `WidgetPlugin`s to render widgets within panels. So if you have a dh.ui panel with a table and plotly-express plot, it needs both as `WidgetPlugin`s to render properly. Just the `DashboardPlugin` won't do anything since that's listening at the dashboard level for a plotly-express widget to be opened as a panel.

vbabich commented:
But it should work with just the `WidgetPlugin`, right? The plan is to phase out `DashboardPlugin` once we add support for `WidgetPlugin` in DHE.

mattrunyon commented:
Yes once we support `WidgetPlugin`s auto opening like in DHC, we'll only need the `WidgetPlugin`. I think we'll need both in the interim in DHE so that outside of dh.ui the `DashboardPlugin` will open the panel, but in dh.ui, the `WidgetPlugin` is needed so dh.ui knows how to render the widget
